File: src/config/config_file_reader.cpp
#include "config/config_file_reader.h"

namespace config
{
    /**
     * Loads the configuration parameters of a black box from the given YAML file
     *
     * @param config_file_name absolute path of a config file
     */
    ComponentMonitorConfig ComponentMonitorConfigFileReader::load(const std::string root_dir, const std::string config_file_name)
    {
        ComponentMonitorConfig params;
        YAML::Node root;
        try
        {
            root = YAML::LoadFile(root_dir + "/" + config_file_name);
        }
        catch (const std::exception& e)
        {
            std::cout << e.what() << "\n";
            return params;
        }

        if (root["name"])
        {
            params.name = root["name"].as<std::string>();
        }
        else
        {
            throw ComponentMonitorConfigException("name not specified");
        }

        if (root["modes"])
        {
            for (auto monitor_config : root["modes"])
            {
                std::string monitor_config_file = monitor_config.as<std::string>();
                MonitorModeConfig mode_config = ComponentMonitorConfigFileReader::loadModeConfig(root_dir, monitor_config_file);
                params.modes.push_back(mode_config);
            }
        }
        else
        {
            throw ComponentMonitorConfigException("modes not specified");
        }

        // if (root["dependencies"])
        // {
        //     for (auto monitor_config : root["dependencies"])
        //     {
        //
        //     }
        // }

        return params;
    }

    MonitorModeConfig ComponentMonitorConfigFileReader::loadModeConfig(const std::string root_dir, const std::string config_file_name)
    {
        MonitorModeConfig params;
        YAML::Node root;
        try
        {
            root = YAML::LoadFile(root_dir + "/" + config_file_name);
        }
        catch (const std::exception& e)
        {
            std::cout << e.what() << "\n";
            return params;
        }

        if (root["name"])
        {
            params.name = root["name"].as<std::string>();
        }
        else
        {
            throw ComponentMonitorConfigException("mode_config: name not specified");
        }

        if (root["mappings"])
        {
            for (auto mapping_it=root["mappings"].begin(); mapping_it != root["mappings"].end(); ++mapping_it)
            {
                YAML::Node params_node = mapping_it->begin()->second;
                FunctionalMappingConfig fn_mapping_params;

                fn_mapping_params.inputs = params_node["inputs"].as<std::vector<std::string>>();
                for (auto output_it=params_node["outputs"].begin(); output_it != params_node["outputs"].end(); ++output_it)
                {
                    YAML::Node output_params_node = output_it->begin()->second;
                    OutputConfig output_params;
                    output_params.name = output_params_node["name"].as<std::string>();
                    output_params.obtained_value_type = output_params_node["type"].as<std::string>();
                    if (output_params_node["expected"])
                    {
                        output_params.expected_value = output_params_node["expected"].as<std::string>();
                    }
                    fn_mapping_params.outputs.push_back(output_params);
                }

                params.mappings.push_back(fn_mapping_params);
            }
        }
        else
        {
            throw ComponentMonitorConfigException("mode_config: mappings not specified");
        }

        return params;
    }
}

File: component_monitoring/config/config_file_reader.py
from __future__ import print_function
from os.path import join
import yaml
from component_monitoring.config.config_params import ComponentMonitorConfig, MonitorModeConfig, \
                                                      FunctionalMappingConfig, OutputConfig

'''An interface for reading component monitor configuration files

@author Alex Mitrevski, Santosh Thoduka
@contact aleksandar.mitrevski@h-brs.de, santosh.thoduka@h-brs.de
'''
class ComponentMonitorConfigFileReader(object):
    '''Loads the configuration parameters of a component monitor from the given YAML file

    Keyword arguments:
    @param root_dir component monitor configuration file directory
    @param config_file_name absolute path of a config file

    '''
    @staticmethod
    def load(root_dir, config_file_name):
        params = ComponentMonitorConfig()

        file_path = join(root_dir, config_file_name)
        root = ComponentMonitorConfigFileReader.__read_yaml_file(file_path)
        if 'component_name' in root.keys():
            params.component_name = root['component_name']
        else:
            print('{0}: component_name not specified'.format(config_file_name))
            return ComponentMonitorConfig()

        if 'description' in root.keys():
            params.description = root['description']
        else:
            print('{0}: description not specified'.format(config_file_name))
            return ComponentMonitorConfig()

        if 'dependencies' in root.keys():
            params.component_dependencies = root['dependencies']

        if 'recovery_actions' in root.keys():
            params.recovery_actions = root['recovery_actions']

        if 'modes' in root.keys():
            for mode_config_file in root['modes']:
                mode_config = ComponentMonitorConfigFileReader.__load_mode_config(root_dir,
                                                                                  mode_config_file)
                params.modes.append(mode_config)
        else:
            print('{0}: modes not specified'.format(config_file_name))
            return ComponentMonitorConfig()

        return params

    '''Loads the configuration parameters of a component monitor mode from the given YAML file

    Keyword arguments:
    @param root_dir component monitor configuration file directory
    @param config_file_name absolute path of a config file

    '''
    @staticmethod
    def __load_mode_config(root_dir, config_file_name):
        params = MonitorModeConfig()

        file_path = join(root_dir, config_file_name)
        root = ComponentMonitorConfigFileReader.__read_yaml_file(file_path)

        if 'name' in root.keys():
            params.name = root['name']
        else:
            print('{0}: name not specified'.format(config_file_name))
            return MonitorModeConfig()

        if 'description' in root.keys():
            params.description = root['description']
        else:
            print('{0}: description not specified'.format(config_file_name))
            return MonitorModeConfig()

        if 'mappings' in root.keys():
            for mapping in root['mappings']:
                mapping_node = mapping['mapping']

                fn_mapping_params = FunctionalMappingConfig()
                fn_mapping_params.inputs = mapping_node['inputs']

                if 'map_outputs' in mapping_node.keys():
                    fn_mapping_params.map_outputs = mapping_node['map_outputs']

                for output in mapping_node['outputs']:
                    output_node = output['output']

                    output_params = OutputConfig()
                    output_params.name = output_node['name']
                    output_params.obtained_value_type = output_node['type']
                    if 'expected' in output_node.keys():
                        output_params.expected_value = output_node['expected']
                    fn_mapping_params.outputs.append(output_params)
                params.mappings.append(fn_mapping_params)
        else:
            print('{0}: mappings not specified'.format(config_file_name))
            return MonitorModeConfig()

        if 'arguments' in root.keys():
            for argument in root['arguments']:
                argument_node = argument['arg']
                arg_name = argument_node['name']
                arg_value = argument_node['value']
                params.arguments[arg_name] = arg_value

        return params

    @staticmethod
    def __read_yaml_file(file_name):
        file_handle = open(file_name, 'r')
        data = yaml.load(file_handle)
        file_handle.close()
        return data
